### PR TITLE
Handle target specific outputs such as .wasm or .dll.lib

### LINKED ISSUE #4500: cdylib on Windows MSVC places .lib import library only under `target/*/deps` while the associated .dll is placed in both `deps` and its parent directory
On all Windows MSVC, Linux and macOS, cdylib places the dynamic library (such as `.dll`) under `./target/{debug,release}/`. However, on Windows MSVC, the associated `.lib` import library is not placed in there, but in `./target/{debug,release}/deps/`.

Is there any reason Cargo puts the `.lib` under `deps`? The `.lib` file is required by MSVC C/C++ compiler (`cl.exe`) when linking a C program against the `.dll`. It would make more sense if the `.lib` is placed under `./target{debt,release}/` together with `.dll`.

```
PS C:\Users\tatsuya\ffi> rustc --version --verbose
rustc 1.20.0 (f3d6973f4 2017-08-27)
binary: rustc
commit-hash: f3d6973f41a7d1fb83029c9c0ceaf0f5d4fd7208
commit-date: 2017-08-27
host: x86_64-pc-windows-msvc
release: 1.20.0
LLVM version: 4.0

PS C:\Users\tatsuya\ffi> cargo --version
cargo 0.21.0 (5b4b8b2ae 2017-08-12)
```

```
Windows PowerShell
Copyright (C) 2016 Microsoft Corporation. All rights reserved.

PS C:\Users\tatsuya> Import-VisualStudioVars 2017 amd64
**********************************************************************
** Visual Studio 2017 Developer Command Prompt v15.0.26730.12
** Copyright (c) 2017 Microsoft Corporation
**********************************************************************
[vcvarsall.bat] Environment initialized for: 'x64'

PS C:\Users\tatsuya> cd .\ffi

PS C:\Users\tatsuya> cat Cargo.toml
[package]
name = "ffi"
version = "0.1.0"
authors = ["Tatsuya Kawano <...>"]

[lib]
crate-type = ["cdylib"]

[dependencies]

PS C:\Users\tatsuya\ffi> cargo clean

PS C:\Users\tatsuya\ffi> cargo build --release
   Compiling ffi v0.1.0 (file:///C:/Users/tatsuya/ffi)
    Finished release [optimized] target(s) in 6.58 secs

PS C:\Users\tatsuya\ffi> dir .\target\release\

    Directory: C:\Users\tatsuya\ffi\target\release

Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----       2017/09/17     21:25                .fingerprint
d-----       2017/09/17     21:25                build
d-----       2017/09/17     21:25                deps
d-----       2017/09/17     21:25                examples
d-----       2017/09/17     21:25                incremental
d-----       2017/09/17     21:25                native
-a----       2017/09/17     21:25              0 .cargo-lock
-a----       2017/09/17     21:25             77 ffi.d
-a----       2017/09/17     21:25         111616 ffi.dll      <== .dll here

PS C:\Users\tatsuya\ffi> dir .\target\release\deps\

    Directory: C:\Users\tatsuya\ffi\target\release\deps

Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a----       2017/09/17     21:25         111616 ffi.dll        <== .dll here (again)
-a----       2017/09/17     21:25           2171 ffi.dll.exp
-a----       2017/09/17     21:25           3852 ffi.dll.lib    <== .lib here
-a----       2017/09/17     21:25        1069056 ffi.pdb
```

### PR DESCRIPTION
**Pull request by tatsuya6502**
Fixes #4500, #4535

Until now, Cargo does not have any knowledge about target-specific output files. It relies solely on rustc for this sort of information (`rustc --print=file-names ...`).

As a result, Cargo does not place some build artifacts (files) to target/{debug,release} directory. These files include *.wasm for wasm32-unknown-emscripten target and *.dll.lib for *-pc-windows-msvc cdylib target.

This commit will add such knowledge to Cargo so that *.wasm and *.dll.lib will be placed in target/{debug,release} directory.

**EDIT**: I added [a summary of changes](https://github.com/rust-lang/cargo/pull/4570#issuecomment-334433635). Please read it for more details of changes.

**IMPORTANT**
Although I added test cases for both wasm32-unknown-emscripten and *-pc-windows-msvc cdylib targets, ~I did not do manual testing on wasm32-unknown-emscripten target as I do not have an environment with emscripten installed. It will be appreciated if anybody tests this change for wasm32-unknown-emscripten target.~  **EDIT**: Tested for both wasm32-unknown-emscripten and x86_64-pc-windows-msvc. Thanks @Herschel for the help.

### CODE REVIEW

**Comment by tatsuya6502** [comment]
> Although I added test cases for both wasm32-unknown-emscripten and *-pc-windows-msvc cdylib targets, I did not do manual testing on wasm32-unknown-emscripten target as I do not have an environment with emscripten installed.

Please do not merge this; we found a problem with wasm target. See [this comment](https://github.com/rust-lang/cargo/issues/4535#issuecomment-334060734) for details.

**Comment by tatsuya6502** [comment]
> Please do not merge this; we found a problem with wasm target. See this comment for details.

I added [one more commit](https://github.com/rust-lang/cargo/pull/4570/commits/f9a541b5f93e66bc5e02c67ecd585d87fb1185cb) to address the problem. 

Now this PR is *ready* for merge.

**Comment by tatsuya6502** [comment]
## Summary of Changes

### For bin target of wasm32-unknown-emscripten

- Copy .wasm file from dep to its parent directory.
  * If crate name contains hyphens (e.g. "wasm-test"), rustc/emcc will generate js with hyphens ("wasm-test.js") and wasm with underscores ("wasm_test.wasm"). The updated codes is aware of this and will copy the correct wasm file.
- Remove metadata suffix from the target js and wasm files in dep. This is required because wasm file name is embedded in generated js file. (See [this comment](https://github.com/rust-lang/cargo/issues/4535#issuecomment-334060734) for details.)

**Before this PR:**

```toml:Cargo.toml
[package]
name = "wasm-test"
# ...
```

```console
% tree target/wasm32-unknown-emscripten/
target/wasm32-unknown-emscripten/
└── debug
    ├── build
    ├── deps
    │   ├── wasm_test-0f5ba9409d843569.js
    │   ├── wasm_test-0f5ba9409d843569.wasm
    │   ├── wasm_test-0f5ba9409d843569.wasm.map
    │   └── wasm_test-0f5ba9409d843569.wast
    ├── examples
    ├── incremental
    ├── native
    ├── wasm-test.d
    └── wasm-test.js

6 directories, 6 files
```

**After this PR:**

```console
% tree target/wasm32-unknown-emscripten/
target/wasm32-unknown-emscripten/
└── debug
    ├── build
    ├── deps
    │   ├── wasm-test.js         <== no metadata, hyphen
    │   ├── wasm_test.wasm       <== no metadata
    │   ├── wasm_test.wasm.map
    │   └── wasm_test.wast
    ├── examples
    ├── incremental
    ├── native
    ├── wasm_test.d
    ├── wasm-test.d
    ├── wasm-test.js
    └── wasm_test.wasm           <== added

6 directories, 8 files
```

### For cdylib/dylib targets of *-pc-windows-msvc

- Copy .dll.lib file from dep to its parent directory.

**Before this PR:**

```toml:Cargo.toml
[package]
name = "cdylib-test"
# ...

[lib]
crate-type = ["cdylib"]
```

```powershell
PS C:\Users\tatsuya\Documents\cdylib-test> dir .\target\debug\deps\

    Directory: C:\Users\tatsuya\Documents\cdylib-test\target\debug\deps

Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a----        10/5/2017  06:20 PM         115200 cdylib_test.dll
-a----        10/5/2017  06:20 PM           2141 cdylib_test.dll.exp
-a----        10/5/2017  06:20 PM           3892 cdylib_test.dll.lib
-a----        10/5/2017  06:20 PM        1085440 cdylib_test.pdb

PS C:\Users\tatsuya\Documents\cdylib-test> dir .\target\debug\

    Directory: C:\Users\tatsuya\Documents\cdylib-test\target\debug

Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----        10/5/2017  06:20 PM                .fingerprint
d-----        10/5/2017  06:20 PM                build
d-----        10/5/2017  06:20 PM                deps
d-----        10/5/2017  06:20 PM                examples
d-----        10/5/2017  06:20 PM                incremental
d-----        10/5/2017  06:20 PM                native
-a----        10/5/2017  06:20 PM              0 .cargo-lock
-a----        10/5/2017  06:20 PM            119 cdylib_test.d
-a----        10/5/2017  06:20 PM         115200 cdylib_test.dll
```

**After this PR:**

```
PS C:\Users\tatsuya\Documents\cdylib-test> dir .\target\debug\deps\

    Directory: C:\Users\tatsuya\Documents\cdylib-test\target\debug\deps


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a----        10/5/2017  06:28 PM         115200 cdylib_test.dll
-a----        10/5/2017  06:28 PM           2141 cdylib_test.dll.exp
-a----        10/5/2017  06:28 PM           3892 cdylib_test.dll.lib
-a----        10/5/2017  06:28 PM        1085440 cdylib_test.pdb


PS C:\Users\tatsuya\Documents\cdylib-test> dir .\target\debug\

    Directory: C:\Users\tatsuya\Documents\cdylib-test\target\debug


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----        10/5/2017  06:28 PM                .fingerprint
d-----        10/5/2017  06:28 PM                build
d-----        10/5/2017  06:28 PM                deps
d-----        10/5/2017  06:28 PM                examples
d-----        10/5/2017  06:28 PM                incremental
d-----        10/5/2017  06:28 PM                native
-a----        10/5/2017  06:28 PM              0 .cargo-lock
-a----        10/5/2017  06:28 PM            119 cdylib_test.d
-a----        10/5/2017  06:28 PM         115200 cdylib_test.dll
-a----        10/5/2017  06:28 PM            123 cdylib_test.dll.d
-a----        10/5/2017  06:28 PM           3892 cdylib_test.dll.lib  <== added
```

File: src/cargo/ops/cargo_rustc/context.rs [hunk 1]
```diff
@@ -478,9 +481,11 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
         // doing this eventually.
         let __cargo_default_lib_metadata = env::var("__CARGO_DEFAULT_LIB_METADATA");
         if !unit.profile.test &&
-            (unit.target.is_dylib() || unit.target.is_cdylib()) &&
+            (unit.target.is_dylib() || unit.target.is_cdylib() ||
+                 (unit.target.is_bin() && self.target_triple().starts_with("wasm32-"))) &&
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_rustc/context.rs]: Could this be refactored into a dedicated method? Something like:

```rust
fn fails_to_work_with_mangled_filename(...) -> bool {
    // ...
}
```

(just to make it a bit more clear as to the intent)

**Comment by tatsuya6502** [comment]
Thank you for reviewing the codes.

>  If you're extra enterprising (this is optional) I think we could tackle #4056 with this infrastructure as well

Ah, alright, there is another. Sure. I will take that. I have macOS environment and will have some time to spare in coming weekend.

**Comment by alexcrichton** [comment]
Nice! My thinking is that we probably just need to lift up the `.dSYM` directory whenever debuginfo is enabled for a profile. Additionally we *may* need to disable mangling entirely, although I'm not sure if that's still required.

That actually reminds me, maybe we should lift up the pdb files produced with MSVC debug info as well?

Again though, both of these are optional, we can always tackle in future PRs :)

**Comment by tatsuya6502** [comment]
I did not have enough time to update this PR last weekend. I was only able to reproduce the debugger problems on macOS and Windows MSVC; no time for updating the codes. I will try to work on them in tomorrow evening (It is Monday evening now in my time zone).

My impression is I can address macOS debugger problem in this PR in a couple of days, but not Windows one. I found the full-path to the PDB file is embedded in Windows binary, so I can't move the PDB to anywhere else. It seems there is a way to remove the full-path via a linker flag. But I guess we will need to change `rustc`'s linker driver to achieve this(?)

**Comment by alexcrichton** [comment]
Ok no worries! Let's go ahead and get this in and iterate in-tree

@bors: r+

**Pull request merged**